_posts/2023-07-12-React-unity-BDD.markdown

```diff
@@ -0,0 +1,560 @@
+
+```tsx
+// addScene.steps.tsx
+
+import { screen, waitFor } from "@testing-library/dom";
+import { defineFeature, loadFeature } from "jest-cucumber";
+import { act } from "react-dom/test-utils";
+
+export const ViewToRender = () =>
+  render(
+    <UnityContextProvider>
+      <AnimationProvider>
+        <AnimationView />
+      </AnimationProvider>
+    </UnityContextProvider>,
+  );
+
+// --------------------------
+// Feature
+// --------------------------
+const feature = loadFeature("./features/addScene.feature");
+
+defineFeature(feature, (test) => {
+  beforeEach(() => {
+    ViewToRender();
+  });
+  test("Add New Scene", ({ given, when, and, then }) => {
+    given("Drew views the animation details", async () => {
+      // wait to see the animation details
+      getAnimationView();
+    });
+    when("Drew adds a new scene", async () => {
+      // Add new scene that will trigger a unity event
+      await addNewScene();
+    });
+    then("The scene has to be added", async () => {
+      // await to see the app state has been updated with new scene
+      await waitFor(() => {
+        const sceneCards = screen.queryAllByTestId(/scene-title-/);
+        expect(sceneCards).toHaveLength(1);
+      });
+    });
+  });
+});
+```
+
+If you execute the test, you will observe a failure because there is currently no corresponding code implemented for it. To enhance the readability of our test steps, we have incorporated helper functions that encapsulate the actions and assertions involved. By doing so, we make the test steps more expressive and easier to understand.
+
+Now, it is time to implement the helper functions and determine the approach for evaluating or mocking the Unity events. If our objective is to verify that the correct command is executed and submitted to Unity upon a user event, such as "add a new scene," we can utilize Jest mocks. However, if we need to mock the callback function when React sends a command to Unity, we must appropriately mock the event handler with mock data. I will illustrate how to accomplish this shortly.
+
+```ts
+// featureFunctions.ts
+export async function selectAnimation(index: number) {
+  await waitFor(() => {
+    expect(screen.queryByTestId(`open-animation-${index}`)).not.toBeNull();
+  });
+
+  act(() => {
+    fireEvent.click(screen.queryByTestId(`open-animation-${index}`));
+  });
+}
+// Mock the Unity Response `returnedPlayerPositionAndPerspective`
+// So that callback is called and the scene is added
+
+async function getPlayerPositionUnityEvent(x: number, y: number, z: number) {
+  const event = JSON.stringify({
+    eventName: "receivedAnimationSceneDetails",
+    args: JSON.stringify({
+      playerPosition: { x: x, y: y, z: z },
+      screenshotData: "base64 image data",
+    }),
+  });
+  act(() => {
+    const handler = eventHandler(true, "");
```

Comment:
not sure why `true` is the first argument and `empty string` is the second value. This needs to be explained.